ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java

```diff
@@ -0,0 +1,312 @@
+        isSecure = clientConfig.getPropertyAsBoolean(CommonClientConfigKey.IsSecure, false);
+        prioritizeVipAddressBasedServers = clientConfig.getPropertyAsBoolean(CommonClientConfigKey.PrioritizeVipAddressBasedServers, prioritizeVipAddressBasedServers);
+        datacenter = ConfigurationManager.getDeploymentContext().getDeploymentDatacenter();
+
+        shouldUseIpAddr = clientConfig.getPropertyAsBoolean(CommonClientConfigKey.UseIPAddrForServer, DefaultClientConfigImpl.DEFAULT_USEIPADDRESS_FOR_SERVER);
+
+        // override client configuration and use client-defined port
+        if (clientConfig.getPropertyAsBoolean(CommonClientConfigKey.ForceClientPortConfiguration, false)) {
+            if (isSecure) {
+                if (clientConfig.containsProperty(CommonClientConfigKey.SecurePort)) {
+                    overridePort = clientConfig.getPropertyAsInteger(CommonClientConfigKey.SecurePort, DefaultClientConfigImpl.DEFAULT_PORT);
+                    shouldUseOverridePort = true;
+                } else {
+                    logger.warn(clientName + " set to force client port but no secure port is set, so ignoring");
+                }
+            } else {
+                if (clientConfig.containsProperty(CommonClientConfigKey.Port)) {
+                    overridePort = clientConfig.getPropertyAsInteger(CommonClientConfigKey.Port, DefaultClientConfigImpl.DEFAULT_PORT);
+                    shouldUseOverridePort = true;
+                } else {
+                    logger.warn(clientName + " set to force client port but no port is set, so ignoring");
+                }
+            }
+        }
+    }
+
+    @Override
+    public List<DiscoveryEnabledServer> getInitialListOfServers() {
+        return obtainServersViaEureka2();
+    }
+
+    @Override
+    public List<DiscoveryEnabledServer> getUpdatedListOfServers() {
+        return obtainServersViaEureka2();
+    }
+
+    public void shutdown() {
+        ServerListReader current = serverListReaderRef.getAndSet(null);
+        if (current != null) {
+            current.shutdown();
+        }
+    }
+
+    /**
+     * TODO prioritizeVipAddressBasedServers not supported yet; all vips are subscribed to eagerly
+     */
+    private List<DiscoveryEnabledServer> obtainServersViaEureka2() {
+        logger.info("Resolving {} from Eureka2...", this.vipAddresses);
+        List<InstanceInfo> instanceInfos = getServerListReader().getLatestServerListOrWait();
+        if (instanceInfos == null) {
+            logger.warn("Server resolve for vip={},secure={} timed out", this.vipAddresses, isSecure);
+            return Collections.emptyList();
+        }
+        return toServerList(instanceInfos);
+    }
+
+    private ServerListReader getServerListReader() {
+        ServerListReader current = serverListReaderRef.get();
+        if (current != null && current.isClosed()) {
+            serverListReaderRef.compareAndSet(current, null);
```

Comment:
Yes, it is correct. We want to set it to null, only if current object is there. As this method may be run concurrently, another thread may initialize new serverListReader at the same time.